README.md
# 敬请期待
## 从主站获取的最新种子(所有)和所有种子的url对应的名字将在元旦上线
#3 可以根据自己的兴趣过滤喜欢的视频，excel玩的溜是必须的
# 91 介绍与获取
[91 porn](http://91porn.com/) 是一个知名的自拍视频网站
- 本程序获取所有视频url-
- 然后将视频的下载地址保存到文件，每个文件1000个，可以直接拖到迅雷里下载，一晚上可以下载<label style="color:red"> 300~400GB </label> 没有任何问题
- 可以输入指定的页码进行抓取

# 安装
- [python3](http://www.python.org)
- [redis](redis.io)

On Windows

    redis-server.exe redis.window.conf    

Or on Unix-like OS

    $ redis-server redis.conf 

安装 requests, 快速获取html

    $ pip install requests    

安装 pyquery 解析视频源地址时用的到

    $ pip install pyquery

# 启动

    python run.py    

或者

### 双击  <label style="color:red">**run.bat**</label>运行

## 说明
- parse.bat 解析Redis中的列表，获取视频的具体地址
- run.bat 获取所有视频列表，具体视频地址由parse.bat批处理文件解析
- src2file.py 将所有视频source复制到文本中，将文本的内容 `Ctrl` + `C`, 然后`Ctrl` + `V` 复制到迅雷里疯狂的下载吧

# 忠告
  爱惜自己

# 免责声明
  本程序仅做学习交流之使用，如有其它用途并产生其它后果，本人概不负责。


CHANGELOG
## Change log

### [v3.0] - 2018-01-01
####  新增特性 
- 所有种子从主站获取
- 兼容了所有操作系统，以前只适配windows 

#### 修复 & 优化
- 修复redis中key的存储不合理
- 提高了抓取速度

### [v2.0] - 2017-12-22 
#### 新增特性
- 上传了备站所有种子
- 重命名了批处理脚本

### [v1.0] - 2017-10-01
- 发布第一版，可以从备站正常获取

#### 不足
只适配windows， 后续版本会适配所有系统


download.py
import urllib.request as request
import random, redis, threading
import ctypes
import os
import platform
import sys

client = redis.StrictRedis("localhost", 6379)

def disk(folder):
    if platform.system() == 'Windows':
        free_bytes = ctypes.c_ulonglong(0)
        ctypes.windll.kernel32.GetDiskFreeSpaceExW(ctypes.c_wchar_p(folder), None, None, ctypes.pointer(free_bytes))
        return free_bytes.value/1024/1024/1024 
    else:
        st = os.statvfs(folder)
        return st.f_bavail * st.f_frsize/1024/1024

def download(url):
    randIP = str(random.randint(0, 255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255))
    req = request.Request(url)
    req.add_header('User-Agent', "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:53.0) Gecko/20100101 Firefox/53.0")
    req.add_header('X-Forwarded-For', randIP)
    response = request.urlopen(req)
    file_size = int(response.getheader("Content-Length"))
    bytes_received = 0
    dir = "f:/"

    try:
        with open(dir + str(random.randint(1, 99999999999999999999)) + ".mp4", 'wb') as dst_file:
            while bytes_received / file_size != 1:
                _buffer = response.read(1024 * 1024)

                bytes_received += len(_buffer)
                dst_file.write(_buffer)
                print(threading.current_thread().name + " 已下载 " +  str(bytes_received / file_size))

    except KeyboardInterrupt:
        raise KeyboardInterrupt(
            "Interrupt signal given. Deleting incomplete video.")

def enter(**kwargs):
	start = kwargs["start"]
	end = kwargs["end"]

	for t in range(start, end):
		lst = client.lrange("91_src", start, end)
		for a in lst:
			src = a.decode("utf-8")
			download(src)
			print(threading.current_thread().name, " 下载 ", src, " 完成， 从redis 删除")
			client.lrem("91_src", 1, src)

if __name__ == "__main__":
	# thread_list = []

	# for i in range(1, 6):
	# 	start = (i - 1) * 4000 + 1
	# 	end = i * 4000 + 1
	# 	t = threading.Thread(target=enter, name="a" + str(i),kwargs={'start':start, 'end':end})
	# 	thread_list.append(t)

	# for t in thread_list:
	# 	t.start()

	# for t in thread_list:
	# 	t.join()

	# print("over")
	# enter(start=1, end=2)


parse_src.py
import requests, re, redis, redisutil, time, random
from pyquery import PyQuery as pq 
from urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter
import threading
import common

# 将列表页插入redis
def parse(url, c):
    html = common.visit1(url)
    if html == "error":
        redisutil.add(url, "parse_error")
        print("parsing url ", url, " has some errors")
        return
    d = pq(html)
    src = d("video").find("source").attr("src")
    title = d("head").find("title").html()
   	# 解析其他信息，时长，名字等
    if src != None:
    	print( threading.current_thread().name,  " insert into redis ", src)
    	c.lrem(common.KEY, 1, url)  # KEY 存储即将解析的url, VISITED 存储已经访问过并且已经 解析的url
    	c.rpush(common.VISITED, url)
    	c.rpush(common.SRC, src)
    	print(threading.current_thread().name, " 解析了 ", title)
    	if src is not None:
    		name = src.split("?")[0].split("/")[-1]    	
    		c.hset("91_detail", name, title) # 文件名和中文名的对应关系
    else:
    	print(threading.current_thread().name,  src, "解析为None, 插入 redis_error")
    	redisutil.add(src, common.KEY_NONE)

def enter(**kwargs):
    start = kwargs["start"]
    end = kwargs["end"]
    c = redisutil.connect()
    lst = c.lrange(common.KEY, int(start), int(end))

    for a in lst:
         print(threading.current_thread().name,  " parsing url ", a)
         parse(a, c)
         time.sleep(0.1)
    with open(common.PARSE_LOG, "a") as f:
        f.write(threading.current_thread().name + " 已经解析完毕.\n")

def start():
    thread_list = []
    total = redisutil.total(common.KEY   )
    page_size = 0
    thread_total = 5

    if total <= 5:
    	page_size = 1
    	thread_total = total
    else:
    	page_size = total / 5

    for t in range(1, thread_total + 1):
        start = (t - 1) * page_size + 1
        end = t * page_size + 1
        name = "a" + str(t)
        t = threading.Thread(target=enter, name=name, kwargs={"start":start, "end":end})
        thread_list.append(t)

    for t in thread_list:
    	t.start()

    for t in thread_list:
    	t.join()

    print("all thread over")

start()


parse_list.py
import requests, re, redisutil, time, random, threading, os
from urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter
import common

# 将列表页插入redis
def parseList(url, page):
	# 将当前现场访问的页码写入日志
    with open("log/" + threading.current_thread().name + ".log", "w") as f:
        f.write(str(page))

	# 程序强制退出后下次就不再从第一页开始了，直接从文件中读取
    html = common.visit(url)
    if html == "error" :
        redisutil.add(url, "error")
        print("出现错误了, ", url)
        return
    lst = re.compile(r'href="(.+?)"').findall(html)
    result = []
    for a in lst:
        if "category=mr" in a:
            result.append(a)
    if result is not None and len(result) > 0:
        for a in set(result): 
            if not redisutil.exists(a, common.KEY) and not redisutil.exists(a, common.VISITED):
                redisutil.add(a, common.KEY)    

                print(threading.current_thread().name, " insert into redis ", a)
            else:
                print(threading.current_thread().name, " redis 已经存在，不再访问 ", a)

'''
    线程主方法 
'''
def enter(**kwargs):
    start = kwargs["start"]
    end = kwargs["end"]
    for page in range(start, end):
        url = common.URL + "/v.php?next=watch&page=" + str(page)
        try:
            print(threading.current_thread().name, " 解析 ", page, " 页 ", url)
            # print(threading.current_thread().name, " is invoking this method")
            parseList(url, page)
            # print(threading.current_thread().name, " finished invoking this method")
            time.sleep(random.randint(1,3))
        except RuntimeError:
            print(threading.current_thread().name, " visiting page ", page, " occurs some errors ", RuntimeError.__with_traceback__)
            redisutil.add(url, "91_error")
            continue
    # current thread has finished, log it and we can easily know it 
    with open(common.LOG, "a") as f:
    	f.write("线程" + threading.current_thread().name + " 已经完成抓取 \n")

# 运行方法
def start():
    thread_list = []
    total = common.getNumber()
    thread_total = 5 # 线程总数，默认为5，如果抓取页面小于5，则线程总数就是抓取的页面总数

    if total <= 5:
        page_size = 1
        thread_total = total
    else:
        page_size = int(total / 5) # start 5 thread to visit
    print("start ", thread_total, " thread to get data")


    for i in range(1, thread_total + 1):
        file = "log/" + str(i) + ".log"
        pre = None

        if os.path.exists(file):
            pre = open(file).read()
        
        start = 0

        if pre is None or pre == "":
            start = (i - 1) * page_size + 1
        else:
            start = int(pre)

        end = i * page_size + 1
        name = "a" + str(i)
        t = threading.Thread(target=enter, name=name, kwargs={"start":start,"end":end})
        thread_list.append(t)

    for t in thread_list:
        t.start()

    for t in thread_list:
        t.join()
		
    print("all thread over")

start()


common.py
import requests, re, redisutil, time, random, threading
from urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter
import urllib.request as request

cookies = requests.cookies.RequestsCookieJar()
cookies.set("language", "cn_CN", domain=".91.91p17.space", path="/")

#--------------------------------------
# 91 的临时站点，可以随时更换
URL = "http://91porn.com"
KEY = "91"
KEY_NONE = "91_none"
LOG = "log/visit.log"
TORRENT = "sed/"
PARSE_LOG = "log/parse.log"
VISITED = "91_visited"
SRC = "91_src"
#----------------------------------------
import os
path = "/".join(LOG.split("/")[0:-1])

print(path)

if not os.path.exists(TORRENT):
	os.makedirs(TORRENT)

if not os.path.exists(path):
    os.makedirs(path)


'''
  获取访问的主页面
'''
def getNumber():
    r = 0
    while True:
        num = input("请输入你想抓取的总页数:")
        try:
            r = int(num)
            break
        except:
            print("抱歉，您输入的不是有效的数字, 请重新输入.")
            continue
    return r


'''
   构造随机ip作为请求头访问目标站点
'''
def visit1(url):
    try:
        randomIP = str(random.randint(0, 255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255))
        retries = Retry(total=5,backoff_factor=10, status_forcelist=[500,502,503,504])
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:53.0) Gecko/20100101 Firefox/53.0',
            'X-Forwarded-For': randomIP}
        s = requests.Session()
        s.mount('http://', HTTPAdapter(max_retries=retries))
        html = s.get(url, headers=headers, cookies=cookies, timeout=5).text #设置比5秒超时，如果可以正常访问，则timeout不起作用，否则达到timeout之后，就可以出异常了，然后返回错误，，程序就不会一直死在那里了
        return html
    except:
        return "error"

def visit(url):
    try:

        randomIP = str(random.randint(0, 255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255)) + "." + str(random.randint(0,255))
        # s = requests.Session()
        # s.mount('http://', HTTPAdapter(max_retries=retries))
        # html = s.get(url, headers=headers, cookies=cookies).text
        r = request.Request(url)

        r.add_header("User-Agent", "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:53.0) Gecko/20100101 Firefox/53.0")
        r.add_header("X-Forwarded-For", randomIP)
        html = request.urlopen(r, timeout=5).read().decode("UTF-8")
        return html
    except:
        return "error"
